include/ginkgo/core/distributed/preconditioner/schwarz.hpp

```diff
@@ -94,7 +94,8 @@ class Schwarz
         /**
          * Local solver factory.
          */
-        GKO_DEFERRED_FACTORY_PARAMETER(local_solver, LinOpFactory);
+        std::shared_ptr<const LinOpFactory> GKO_DEFERRED_FACTORY_PARAMETER(
```

Comment:
Do we need to go back to this pattern? The macro becomes much more complex due to all the `decltype`s, I deliberately did not go this way with the new macros.

Reply:
Both are fine with me. I think it is good because it gives the same pattern like the other macro.
You only mean the decltype of the distinct parameter not the `decltype(self())` for the reference return, right?
`decltype(self())` is necessary to avoid additional input argument of macro or assumption of the type name.

Reply:
The `_type` parameter in the original `GKO_DEFERRED_FACTORY_PARAMETER` macro occurs 4 times. We need to replace all of those by `decltype(_name_)`, which complicates the macro.

Reply:
No, it only need once. we have `using _name_type = ...` and then all `_type` just use that.
`mutable` is not longer there as we said it is not intended behavior in the #1336 
changing the member and using `with_*` are only available in non-const object.

Reply:
This solution is still more complex than leaving the type as a macro parameter, you need to use `_name##_type` instead of `_type` everywhere and add the type alias.

Reply:
For me the main reason to use this style is that it's consistent with the non-deferred parameters. I think Tobias raises a good point that the other style makes it clearer that all the code gets generated. It might also help to document the functions, although we still might not be able to rely on generated documentation.
But for now I would prefer this style for consistency, and then with an interface break change to the other style for all parameters.